README.md
# cr-easeshow

react component

## Example
<img src="./github/easeshow2.gif" width="800px" alt="example" />

## Usage

```jsx
cnpm install --save cr-easeshow
```

```js
import EaseShow from 'cr-easeshow'
import 'cr-easeshow/assets/index.css'
// 注意在使用的时候不能忘记引入css
<div style={{height: 600, border: '1px solid #21c8be'}}>
  <EaseShow minScale={0.5} maxScale={2}>
    <div style={{fontSize: 200, color: '#e850e6'}}>HELLO WORLD</div>
  </EaseShow>
</div>
```

## 注意
如果使用了cssModule，作为样式的使用方式，在引入css时应特别注意
以And Design Pro使用为例，在对应的页面less中引入css。注意包名前面的波浪线
```less
@import '~cr-easeshow/assets/index.css';
```
## API
name | type | description
-----|------|------------
minScale|number|缩小的最小倍数
maxScale|number|放大的最大倍数
bgColor|string|拖动放大缩小画布的背景颜色
scale|number|组件可以按照外部值放大缩小
isMove|bool|组件是否可以拖动`true`可以拖动
isScale|bool|组件是否可以缩放`true`可以缩放

## Development

```
cnpm install
npm start
```

## Example

http://localhost:8021/examples/testOne.html


## License

cr-easeshow is released under the MIT license.


HISTORY.md
# History
----

## 1.0.0 / 2019-06-03

cr-easeshow is start

## 1.0.1 / 2019-06-13
增加scale接口，组件可以按照外部值放大缩小

## 1.0.2 / 2019-06-14
增加isMove接口
增加isScale接口




src/index.js
import React, { PureComponent, Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import OuiDom from './utils/ouiDomUtils'
import './utils/SimpleDrag'
import isEqual from 'lodash/isEqual'

const showW = 99999
const showH = 99999
let scale = 1

function moveTo(ele, offsetx, offsety) {
  const positionProperty = ['static', 'relative', 'absolute', 'sticky', 'fixed']
  const eleStyle = OuiDom.getComputedStyle(ele, 'position')
  if (eleStyle === 'static' || eleStyle === 'sticky') {
    ele.style.position = 'relative'
  }
  ele.style.left = offsetx + 'px'
  ele.style.top = offsety + 'px'
}

class EaseShow extends PureComponent {

  constructor(props) {
    super(props)
    this.state = {
      minScale: props.minScale,
      maxScale: props.maxScale,
      bgColor: props.bgColor,      
    }
  }

  componentDidMount() {
    const drag = document.getElementById('drag')
    const showBox = this.showBox
    const outBox = this.outBox
    const container = this.container
    const { isMove } = this.props
    const propScale = this.props.scale
    scale = propScale
    OuiDom.setStyles(showBox, {
      ['width']: showW + 'px',
      ['height']: showH + 'px',
      ['background-color']: this.state.bgColor,
    })

    const boxW = OuiDom.outerWidth(outBox)
    const boxH = OuiDom.outerHeight(outBox)
    OuiDom.setStyles(container, {
      ['width']: boxW + 'px',
      ['height']: boxH + 'px',    
    })
    this.setScale(showBox, propScale)

    this.setMove(outBox, showBox)
    
    if (isMove) {
      drag.sdrag()
    }
  }
  componentDidUpdate(prevProps, prevState) {
    // console.log('component up date', prevState)
    if (!isEqual(this.props, prevProps)) {
      const { scale } = this.props
      this.setScale(this.showBox, scale)
    }
  }

  setMove = ( wind, canvas ) => {
    let startX = 0
    let startY = 0
    const dragW = OuiDom.outerWidth(canvas)
    const dragH = OuiDom.outerHeight(canvas)
    const conW = OuiDom.outerWidth(wind)
    const conH = OuiDom.outerHeight(wind)
    startX = (conW - dragW) / 2
    startY = (conH - dragH) / 2
    // console.log('moveTo----', startX, startY)
    moveTo(canvas, startX, startY)
  }

  setScale = (ele, scale) => {
    scale = Math.min(Math.max(this.state.minScale, scale), this.state.maxScale);
    if (ele) {
      ele.style.transform = `scale(${scale})`;
    }
  }

  handleWheel = (event) => {
    const { isScale } = this.props
    if (!isScale) {
      return
    }
    // const ele = event.target
    const ele = this.showBox
    if (event.deltaY > 0) {
      scale += 50 * -0.001;
    } else {
      scale += 50 * 0.001;
    }
  
    // Restrict scale
    scale = Math.min(Math.max(this.state.minScale, scale), this.state.maxScale);
    console.log(scale)
  
    // Apply scale transform
    ele.style.transform = `scale(${scale})`;
    ele.style.transformOrigin = 'center'
    // console.log(scale)   
  }

  render () {
    const { prefixCls } = this.props
    
    return (
      <div className={`${prefixCls}`} ref={outBox => this.outBox = outBox}>
        <div id='drag' className={`${prefixCls}-showBox`} ref={showBox => this.showBox = showBox} onWheel={this.handleWheel}>
          <div className={`${prefixCls}-container`} ref={container => this.container = container}>
            {this.props.children}
          </div>
        </div>
      </div>
    )
  }
}

EaseShow.propTypes = {
  minScale: PropTypes.number,
  maxScale: PropTypes.number,
  bgColor: PropTypes.string,
  prefixCls: PropTypes.string,
  scale: PropTypes.number,
  isMove: PropTypes.bool,
  isScale: PropTypes.bool,
}
EaseShow.defaultProps = {
  minScale: 0.125,
  maxScale: 3,
  bgColor: '#fae0dd',
  prefixCls: 'cr-easeshow',
  isMove: true,
  isScale: true,
}
export default EaseShow

src/utils/SimpleDrag.js
(function () {

  /**
   * THIS OBJECT WILL ONLY WORK IF your target is positioned relative or absolute,
   * or anything that works with the top and left css properties (not static).
   *
   * Howto
   * ============
   *
   * document.getElementById('my_target').sdrag();
   *
   * onDrag, onStop
   * -------------------
   * document.getElementById('my_target').sdrag(onDrag, null);
   * document.getElementById('my_target').sdrag(null, onStop);
   * document.getElementById('my_target').sdrag(onDrag, onStop);
   *
   * Both onDrag and onStop callback take the following arguments:
   *
   * - el, the currentTarget element (#my_target in the above examples)
   * - pageX: the mouse event's pageX property (horizontal position of the mouse compared to the viewport)
   * - startX: the distance from the element's left property to the horizontal mouse position in the viewport.
   *                  Usually, you don't need to use that property; it is internally used to fix the undesirable
   *                  offset that naturally occurs when you don't drag the element by its top left corner
   *                  (for instance if you drag the element from its center).
   * - pageY: the mouse event's pageX property (horizontal position of the mouse compared to the viewport)
   * - startY: same as startX, but for the vertical axis (and element's top property)
   *
   *
   *
   * The onDrag callback accepts an extra argument: fix.
   *
   * fix is an array used to fix the coordinates applied to the target.
   *
   * It can be used to constrain the movement of the target inside of a virtual rectangle area for instance.
   * Put a variable in the fix array to override it.
   * The possible keys are:
   *
   * - pageX
   * - startX
   * - pageY
   * - startY
   * - skipX
   * - skipY
   *
   * skipX and skipY let you skip the updating of the target's left property.
   * This might be required in some cases where the positioning of the target
   * is automatically done by the means of other css properties.
   *
   * 
   *
   *
   *
   *
   * Direction
   * -------------
   * With direction, you can constrain the drag to one direction only: horizontal or vertical.
   * Accepted values are:
   *
   * - <undefined> (the default)
   * - vertical
   * - horizontal
   *
   *
   *
   *
   */

  // simple drag
  function sdrag(onDrag, onStop, direction) {

      var startX = 0;
      var startY = 0;
      var el = this;
      var dragging = false;

      function move(e) {

          var fix = {};
          onDrag && onDrag(el, e.pageX, startX, e.pageY, startY, fix);
          if ('vertical' !== direction) {
              var pageX = ('pageX' in fix) ? fix.pageX : e.pageX;
              if ('startX' in fix) {
                  startX = fix.startX;
              }
              if (false === ('skipX' in fix)) {
                  el.style.left = (pageX - startX) + 'px';
              }
          }
          if ('horizontal' !== direction) {
              var pageY = ('pageY' in fix) ? fix.pageY : e.pageY;
              if ('startY' in fix) {
                  startY = fix.startY;
              }
              if (false === ('skipY' in fix)) {
                  el.style.top = (pageY - startY) + 'px';
              }
          }
      }

      function startDragging(e) {
          if (e.currentTarget instanceof HTMLElement || e.currentTarget instanceof SVGElement) {
              dragging = true;
              var left = el.style.left ? parseInt(el.style.left) : 0;
              var top = el.style.top ? parseInt(el.style.top) : 0;
              startX = e.pageX - left;
              startY = e.pageY - top;
              window.addEventListener('mousemove', move);
          }
          else {
              throw new Error("Your target must be an html element");
          }
      }

      this.addEventListener('mousedown', startDragging);
      window.addEventListener('mouseup', function (e) {
          if (true === dragging) {
              dragging = false;
              window.removeEventListener('mousemove', move);
              onStop && onStop(el, e.pageX, startX, e.pageY, startY);
          }
      });
  }

  Element.prototype.sdrag = sdrag;
})();

examples/testOne.js
import React, { PureComponent } from 'react'
import ReactDOM from 'react-dom'
import PropTypes from 'prop-types'
import EaseShow from 'cr-easeshow'
import '../assets/index.less'

const reactContainer = document.getElementById('__react-content');
const bodyContainer = document.getElementsByTagName('body')
bodyContainer[0].style.padding = '10px'
bodyContainer[0].style.overflow = 'hidden'

class App extends PureComponent {
  state = {
    scale: 10
  }
  componentDidMount() {
    // setTimeout(() => {
    //   console.log('执行延时')
    //   this.setState({
    //     scale: 2
    //   })
    // }, 1000)
  }
  handleClick = () => {
    console.log('click')
  }
  
  render () {
    return (
      <div>
        <div style={{height: 600, border: '1px solid #21c8be'}}>
          <EaseShow minScale={0.2} maxScale={2} isMove={true} isScale={true} scale={this.state.scale}>
            <div onClick={this.handleClick} style={{fontSize: 200, color: '#e850e6'}}>HELLO WORLD</div>
          </EaseShow>
        </div>
      </div>
    )
  }
}


function render(container) {
  ReactDOM.render(
    <div>
      <h3>EASESHOW TEST</h3>
      <div><App /></div>               
    </div>, container
  )
}

render(reactContainer)